Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -1263,4 +1275,237 @@ lemma nonsingularLift_addMap {P Q : PointClass F} (hP : W.NonsingularLift P)
+            toAffine_add_of_Z_ne_zero hP hQ hPz hQz hxy]
+      · rw [_root_.not_imp, not_ne_iff] at hxy
+        rw [toAffine_of_Z_ne_zero hP hPz, toAffine_of_Z_ne_zero hQ hQz, Affine.Point.add_of_Y_eq
+            ((X_eq_iff hPz hQz).mp hxy.left) ((Y_eq_iff' hPz hQz).mp hxy.right)]
+        by_cases hy : P y * Q z ^ 3 = Q y * P z ^ 3
+        · rw [add_of_Y_eq hPz hQz hxy.left hy hxy.right,
+            toAffine_smul _ <| isUnit_dblU_of_Y_eq hP hPz hQz hxy.left hy hxy.right, toAffine_zero]
+        · rw [add_of_Y_ne hP.left hQ.left hPz hQz hxy.left hy,
+            toAffine_smul _ <| isUnit_addU_of_Y_ne hPz hQz hy, toAffine_zero]
+
+/-- The map from a nonsingular rational point on a Weierstrass curve `W` in Jacobian coordinates
+to the corresponding nonsingular rational point on `W` in affine coordinates. -/
+noncomputable def toAffineLift (P : W.Point) : W.toAffine.Point :=
+  P.point.lift _ fun _ _ => toAffine_of_equiv
+
+lemma toAffineLift_eq {P : Fin 3 → F} (hP : W.NonsingularLift ⟦P⟧) :
+    toAffineLift ⟨hP⟩ = toAffine W P :=
+  rfl
+
+lemma toAffineLift_of_Z_eq_zero {P : Fin 3 → F} (hP : W.NonsingularLift ⟦P⟧) (hPz : P z = 0) :
+    toAffineLift ⟨hP⟩ = 0 :=
+  toAffine_of_Z_eq_zero hP hPz
+
+lemma toAffineLift_zero : toAffineLift (0 : W.Point) = 0 :=
+  toAffine_zero
+
+lemma toAffineLift_of_Z_ne_zero {P : Fin 3 → F} {hP : W.NonsingularLift ⟦P⟧} (hPz : P z ≠ 0) :
+    toAffineLift ⟨hP⟩ = Affine.Point.some ((nonsingular_of_Z_ne_zero hPz).mp hP) :=
+  toAffine_of_Z_ne_zero hP hPz
+
+lemma toAffineLift_some {X Y : F} (h : W.NonsingularLift ⟦![X, Y, 1]⟧) :
+    toAffineLift ⟨h⟩ = Affine.Point.some ((nonsingular_some ..).mp h) :=
+  toAffine_some h
+
+lemma toAffineLift_neg {P : Fin 3 → F} (hP : W.NonsingularLift ⟦P⟧) :
+    toAffineLift (-⟨hP⟩) = -toAffineLift ⟨hP⟩ :=
+  toAffine_neg hP
+
+lemma toAffineLift_add {P Q : Fin 3 → F} (hP : W.NonsingularLift ⟦P⟧) (hQ : W.NonsingularLift ⟦Q⟧) :
+    toAffineLift (⟨hP⟩ + ⟨hQ⟩) = toAffineLift ⟨hP⟩ + toAffineLift ⟨hQ⟩ :=
+  toAffine_add hP hQ
+
+/-- The equivalence between the nonsingular rational points on a Weierstrass curve `W` in Jacobian
+coordinates with the nonsingular rational points on `W` in affine coordinates. -/
+@[simps]
+noncomputable def toAffineAddEquiv : W.Point ≃+ W.toAffine.Point where
+  toFun := toAffineLift
+  invFun := fromAffine
+  left_inv := by
+    rintro @⟨⟨P⟩, hP⟩
+    by_cases hPz : P z = 0
+    · rw [Point.ext_iff, toAffineLift_eq, toAffine_of_Z_eq_zero hP hPz]
+      exact Quotient.eq.mpr <| Setoid.symm <| equiv_zero_of_Z_eq_zero hP hPz
+    · rw [Point.ext_iff, toAffineLift_eq, toAffine_of_Z_ne_zero hP hPz]
+      exact Quotient.eq.mpr <| Setoid.symm <| equiv_some_of_Z_ne_zero hPz
+  right_inv := by
+    rintro (_ | _)
+    · erw [fromAffine_zero, toAffineLift_zero, Affine.Point.zero_def]
+    · rw [fromAffine_some, toAffineLift_some]
+  map_add' := by
+    rintro @⟨⟨_⟩, _⟩ @⟨⟨_⟩, _⟩
```

Comment:
```suggestion
    rintro ⟨_⟩ ⟨_⟩
```
(untested)

Reply:
This doesn't quite work because `toAffineLift_add` needs the `Fin 3 -> F` extracted out, but I don't know why I have this in the first place. If I generalise `toAffineLift_add` to take in a `PointClass F` it works!